common/src/Common/Network.hs

```diff
@@ -208,3 +226,82 @@ renderNodeRef (NodeRef (URI.Authority mUser h mp)) =
     maybe "" ((<> "@") . renderUser) mUser <> URI.unRText h <> maybe "" ((":" <>) . tshow) mp
   where
     renderUser (URI.UserInfo name mPw) = URI.unRText name <> maybe "" ((":" <>) . URI.unRText) mPw
+
+-- | Helper function for compiling pact code to a list of terms
+compileCode :: Text -> Either String [Term Name]
+compileCode = first show . compileExps mkEmptyInfo <=< parseExprs
+
+-- | Parse the balance checking object into a map of account balance changes and
+-- the result from the inner code
+parseWrappedBalanceChecks :: PactValue -> Either Text (Map AccountName Decimal, PactValue)
+parseWrappedBalanceChecks = first ("parseWrappedBalanceChecks: " <>) . \case
+  (PObject (ObjectMap obj)) -> do
+    let lookupErr k = case Map.lookup (FieldKey k) obj of
+          Nothing -> Left $ "Missing key '" <> k <> "' in map: " <> renderCompactText (ObjectMap obj)
+          Just v -> pure v
+    before <- parseAccountBalances =<< lookupErr "before"
+    result <- parseResults =<< lookupErr "results"
+    after <- parseAccountBalances =<< lookupErr "after"
+    pure (Map.unionWith subtract before after, result)
+  v -> Left $ "Unexpected PactValue (expected object): " <> renderCompactText v
+
+-- | Turn the object of account->balance into a map
+parseAccountBalances :: PactValue -> Either Text (Map AccountName Decimal)
+parseAccountBalances = first ("parseAccountBalances: " <>) . \case
+  (PObject (ObjectMap obj)) -> do
+    m <- for (Map.toAscList obj) $ \(FieldKey accountText, pv) -> do
+      bal <- case pv of
+        PLiteral (LDecimal d) -> pure d
+        t -> Left $ "Unexpected PactValue (expected decimal): " <> renderCompactText t
+      acc <- mkAccountName accountText
+      pure (acc, bal)
+    pure $ Map.fromList m
+  v -> Left $ "Unexpected PactValue (expected object): " <> renderCompactText v
+
+-- | Get the last result, as we would under unwrapped deployment.
+parseResults :: PactValue -> Either Text PactValue
+parseResults = first ("parseResults: " <>) . \case
+  PList vec -> pure $ V.last vec
```

Comment:
`V.last` is partial, if we have an empty list this will throw an exception. Maybe...

```haskell
vec ^? _last
```

Reply:
the joys of partial functions which aren't marked `unsafe` :(